From 15d28855ceb49aac963c007b4e2fbed47bcb6750 Mon Sep 17 00:00:00 2001
From: Simo Piiroinen <simo.piiroinen@jollamobile.com>
Date: Thu, 13 Apr 2017 09:53:13 +0300
Subject: [PATCH] [pam_systemd] Add timeout=<seconds> argument. JB#37881

If user@.service takes longer than 25 seconds to start up, the
CreateSession method call made from pam_systemd will timeout
and failure is assumed (even if the user service gets successfully
to started stage, the reply sent after timeout is ignored).

Add sd_bus_call_method_with_timeout() convenience method call function
that is otherwise similar to already existing sd_bus_call_method()
except it allows using custom reply timeouts.

Allow passing of timeout=<seconds> argument to pam_systemd plugin when
longer than default D-Bus method call timeouts are needed.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jollamobile.com>
---
 src/login/pam_systemd.c | 20 ++++++++++++++++----
 1 file changed, 16 insertions(+), 4 deletions(-)

diff --git a/src/login/pam_systemd.c b/src/login/pam_systemd.c
index 5bd7efc3e8..53d66e69f8 100644
--- a/src/login/pam_systemd.c
+++ b/src/login/pam_systemd.c
@@ -52,7 +52,8 @@ static int parse_argv(
                 const char **class,
                 const char **type,
                 const char **desktop,
-                bool *debug) {
+                bool *debug,
+                unsigned *timeout) {
 
         unsigned i;
 
@@ -87,6 +88,11 @@ static int parse_argv(
                         else if (debug)
                                 *debug = k;
 
+                } else if (startswith(argv[i], "timeout=")) {
+                        if (safe_atou(argv[i] + 8, timeout) != 0)
+                                pam_syslog(handle, LOG_WARNING, "Failed to parse timeout= argument, ignoring.");
+                        else if (timeout)
+                                *timeout = (*timeout * USEC_PER_SEC);
                 } else
                         pam_syslog(handle, LOG_WARNING, "Unknown parameter '%s', ignoring", argv[i]);
         }
@@ -689,6 +695,7 @@ _public_ PAM_EXTERN int pam_sm_open_session(
         _cleanup_(user_record_unrefp) UserRecord *ur = NULL;
         int session_fd = -1, existing, r;
         bool debug = false, remote;
+        unsigned  timeout = LOGIN_SLOW_BUS_CALL_TIMEOUT_USEC;
         uint32_t vtnr = 0;
         uid_t original_uid;
 
@@ -699,7 +706,8 @@ _public_ PAM_EXTERN int pam_sm_open_session(
                        &class_pam,
                        &type_pam,
                        &desktop_pam,
-                       &debug) < 0)
+                       &debug,
+                       &timeout) < 0)
                 return PAM_SESSION_ERR;
 
         if (debug)
@@ -917,7 +925,10 @@ _public_ PAM_EXTERN int pam_sm_open_session(
         if (r < 0)
                 return pam_bus_log_create_error(handle, r);
 
-        r = sd_bus_call(bus, m, LOGIN_SLOW_BUS_CALL_TIMEOUT_USEC, &error, &reply);
+        if (timeout != LOGIN_SLOW_BUS_CALL_TIMEOUT_USEC)
+                pam_syslog(handle, LOG_DEBUG, "Using %us D-Bus method call timeout", (unsigned int)(timeout / USEC_PER_SEC));
+
+        r = sd_bus_call(bus, m, timeout, &error, &reply);
         if (r < 0) {
                 if (sd_bus_error_has_name(&error, BUS_ERROR_SESSION_BUSY)) {
                         if (debug)
@@ -1042,7 +1053,8 @@ _public_ PAM_EXTERN int pam_sm_close_session(
                        NULL,
                        NULL,
                        NULL,
-                       &debug) < 0)
+                       &debug,
+                       NULL) < 0)
                 return PAM_SESSION_ERR;
 
         if (debug)
